src/main/java/com/google/api/gax/bundling/PushingBundler.java

```diff
@@ -0,0 +1,221 @@
+    private Collection<BundlingThreshold<E>> thresholds;
+    private ScheduledExecutorService executor;
+    private Duration maxDelay;
+    private ThresholdBundleReceiver<E> receiver;
+    private BundlingFlowController<E> flowController;
+    private BundleMerger<E> bundleMerger;
+
+    private Builder() {}
+
+    public Builder<E> setThresholds(Collection<BundlingThreshold<E>> thresholds) {
+      this.thresholds = thresholds;
+      return this;
+    }
+
+    public Builder<E> setExecutor(ScheduledExecutorService executor) {
+      this.executor = executor;
+      return this;
+    }
+
+    public Builder<E> setMaxDelay(Duration maxDelay) {
+      this.maxDelay = maxDelay;
+      return this;
+    }
+
+    public Builder<E> setReceiver(ThresholdBundleReceiver<E> receiver) {
+      this.receiver = receiver;
+      return this;
+    }
+
+    public Builder<E> setFlowController(BundlingFlowController<E> flowController) {
+      this.flowController = flowController;
+      return this;
+    }
+
+    public Builder<E> setBundleMerger(BundleMerger<E> bundleMerger) {
+      this.bundleMerger = bundleMerger;
+      return this;
+    }
+
+    public PushingBundler<E> build() {
+      return new PushingBundler<>(this);
+    }
+  }
+
+  public static <E> Builder<E> newBuilder() {
+    return new Builder<>();
+  }
+
+  /**
+   * Adds an element to the bundler. If the element causes the collection to go past any of the
+   * thresholds, the bundle will be sent to the {@code ThresholdBundleReceiver}.
+   */
+  public void add(E e) throws FlowControlException {
+    // We need to reserve resources from flowController outside the lock, so that they can be
+    // released by trigger().
+    flowController.reserve(e);
+    lock.lock();
+    try {
+      receiver.validateBundle(e);
+      boolean anyThresholdReached = isAnyThresholdReached(e);
+
+      if (currentOpenBundle == null) {
+        currentOpenBundle = e;
+        if (!anyThresholdReached) {
+          // Do not schedule a job when a threshold is exceeded, as it will be immediately cancelled
+          currentAlarmFuture =
+              executor.schedule(flushRunnable, maxDelay.getMillis(), TimeUnit.MILLISECONDS);
+        }
+      } else {
+        bundleMerger.merge(currentOpenBundle, e);
+      }
+
+      if (anyThresholdReached) {
+        trigger();
+      }
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  public boolean isEmpty() {
+    lock.lock();
+    try {
+      return currentOpenBundle == null;
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  /**
+   * Trigger processing of the next bundle.
+   */
+  public void trigger() {
+    final E bundle = removeBundle();
+    if (bundle != null) {
+      receiver
+          .processBundle(bundle)
+          .addListener(
+              new Runnable() {
+                @Override
+                public void run() {
+                  flowController.release(bundle);
+                }
+              },
```

Comment:
maybe make a member variable holding this Runnable, like the variable currently named `flushRunnable`